javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,187 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.isWebAssembly
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+class ArrayList[E] private (innerInit: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  /* This class has two different implementations for handling the
+   * internal data storage, depending on whether we are on Wasm or JS.
+   * On JS, we utilize `js.Array`. On Wasm, for performance reasons,
+   * we avoid JS interop and use a scala.Array.
+   * The `_size` field (unused in JS) keeps track of the effective size
+   * of the underlying Array for the Wasm implementation.
+   */
+
+  private val innerJS =
+    if (isWebAssembly) null
+    else innerInit.asInstanceOf[js.Array[E]]
+
+  private var innerWasm =
+    if (!isWebAssembly) null
+    else innerInit.asInstanceOf[Array[Any]]
+
   def this(initialCapacity: Int) = {
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly)
+        new Array[Any](Math.max(initialCapacity, 0))
+      else new js.Array[E],
+      0
+    )
     if (initialCapacity < 0)
       throw new IllegalArgumentException
   }
 
   def this() =
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly) new Array[Any](16)
+      else new js.Array[E],
+      0
+    )
 
   def this(c: Collection[_ <: E]) = {
-    this()
+    this(
+      if (isWebAssembly) new Array[Any](c.size())
+      else new js.Array[E],
+      0
+    )
     addAll(c)
   }
 
   def trimToSize(): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly) resizeTo(size())
```

Comment:
Coding style: Always use a separate line for side-effecting branches of control structures (see [here](https://github.com/scala-js/scala-js/blob/main/CODINGSTYLE.md#ifelse))